### DISCUSSION TITLE
Regarding the hyperparameter optimization

### DISCUSSION THREAD

**Discussion by Githubprojectsacc** [ultralytics/ultralytics#10236] [Docs]
HI, Im interested to perform hyperparameter optimization with yolov8 model for the object detection.
I had explored the yolov8 tuning guide and ray tune optimization for performing the tuning process. I used the ASHA schedular with Hyperopt search algorithm for the tuning process with search space defined. I need to get the metrics value to be returned after training the model after each trial. But in this case it shows the same metrics in each trial. I need to get the different metrics for each trial for hyperparameter optimization. Can u pls help me with this..

```
from ultralytics import YOLO
import os
import ray
from ray import tune
from ray import train
from ray.tune import ResultGrid
from ray.train import Result
from ray.tune.schedulers import ASHAScheduler
from ray.tune.search.hyperopt import HyperOptSearch

model = YOLO("yolov8n.pt")
data = '/content/drive/MyDrive/data.yaml'
def model_train(config):
           batch=config["batch_size"]
           learning_rate=config["learning_rate"]
           num_layers=config["num_layers"]
           activation=config["activation"]
           optimizer=config["optimizer"]
           momentum=config["momentum"]
           trained_model = model.train(data=data,
                            batch=batch,
                            epochs=2,
                            workers=8, 
                            imgsz=640, 
                            patience=1,
                            learning_rate=learning_rate,
                            num_layers=num_layers,
                            activation=activation,
                            optimizer=optimizer,
                            momentum=momentum,
                            )
    
          eval=model.val()  
          result=eval.results_dict 
          metric=eval.box.map
          return trained_model.box.map
   
ray.init(ignore_reinit_error=True)

search_space = {
    "learning_rate": tune.loguniform(1e-4, 1e-1),
    "batch_size": tune.choice([8, 16, 32]),
    "num_layers": tune.choice([3, 4, 5]),
    "activation": tune.choice(["relu", "tanh"]),
    "optimizer":tune.choice(['SGD', 'Adam', 'AdamW', 'NAdam', 'RAdam', 'RMSProp']),
    "momentum": tune.loguniform(0.7, 0.98, 0.3)
   
}

search_alg = HyperOptSearch(metric="metric", mode="max")

scheduler = ASHAScheduler(max_t=10, grace_period=1)
trainable_with_resources = tune.with_resources(model_train,{"cpu":1,"gpu":1})
tuner=tune.Tuner(tune.with_resources(    model_train,resources={"cpu":1,"gpu":1}),
    tune_config=tune.TuneConfig(
        metric="metric",
        mode="max",),
        param_space=search_space,
)
result_grid: ResultGrid = tuner.fit()
ray.shutdown()

                          
                                               
                          
    


```

**Comment by glenn-jocher** [comment]
@Githubprojectsacc hey there! 👋 It looks like you've set up most of your hyperparameter tuning process correctly with Ray Tune and Ultralytics YOLOv8. However, I noticed a potential issue in how you're returning metrics from your `model_train` function.

In your `model_train` function, you're attempting to return `trained_model.box.map`, but `trained_model` doesn't directly have the `.box.map` attribute. Instead, you correctly obtain the metric with `metric=eval.box.map` and should return this `metric` at the end of your function.

Here's the corrected return statement:
```python
return {"metric": metric}  # Correctly return the metric in a dictionary
```

Also, ensure that your `tuner` setup's `metric` parameter in `TuneConfig` and `HyperOptSearch` is aligned with what you return from `model_train`. Since we're returning a dictionary like `{"metric": metric}`, it's important that `metric="metric"` is correctly specified, which you've done correctly.

Remember, returning metrics as a dictionary allows Ray Tune to recognize and utilize those metrics for further decision-making in the optimization process.

Hope this helps you get different metrics values for each trial! 😊 Keep explore the incredible possibilities of optimizing YOLOv8 models!
